pkg/api/api.go

```diff
@@ -190,67 +233,105 @@ func (h *APIHandler) Bulk(w http.ResponseWriter, r *http.Request) {
+		} else {
+			actionStatus = h.deterministicHandler(action, b)
+			item = map[string]any{action.Action: map[string]any{"status": actionStatus}}
 		}
-		for k := range j {
-			switch k {
-			case "index":
-				h.metrics.bulkIndexTotalMetrics.Add(context.Background(), 1, attrs)
-				skipNextLine = true
-			case "create":
-				skipNextLine = true
-				actionStatus := h.ActionOdds[rand.Intn(len(h.ActionOdds))]
-				switch actionStatus {
-				case http.StatusOK:
-					h.metrics.bulkCreateOkMetrics.Add(context.Background(), 1, attrs)
-				case http.StatusConflict:
-					br.Errors = true
-					h.metrics.bulkCreateDuplicateMetrics.Add(context.Background(), 1, attrs)
-				case http.StatusTooManyRequests:
-					br.Errors = true
-					h.metrics.bulkCreateTooManyMetrics.Add(context.Background(), 1, attrs)
-				case http.StatusNotAcceptable:
-					br.Errors = true
-					h.metrics.bulkCreateNonIndexMetrics.Add(context.Background(), 1, attrs)
-				}
-				br.Items = append(br.Items, map[string]any{"created": map[string]any{"status": actionStatus}})
-			case "update":
-				h.metrics.bulkUpdateTotalMetrics.Add(context.Background(), 1, attrs)
-				skipNextLine = true
-			case "delete":
-				h.metrics.bulkDeleteTotalMetrics.Add(context.Background(), 1, attrs)
-				skipNextLine = false
-			}
+
+		isErr := h.updateMetrics(action.Action, actionStatus, attrs)
+		br.Errors = br.Errors || isErr
+		if len(item) > 0 {
+			br.Items = append(br.Items, item)
 		}
 	}
+
 	h.recordRequest(r, body)
 	brBytes, err := json.Marshal(br)
 	if err != nil {
 		log.Printf("error marshal bulk reply: %s", err)
 		return
 	}
 	w.Header().Set(http.CanonicalHeaderKey("Content-Type"), "application/json")
-	w.Write(brBytes)
+	_, _ = w.Write(brBytes)
 	return
 }
 
+func (h *APIHandler) parseAction(b []byte) (Action, error) {
+	var j map[string]json.RawMessage
+	err := json.Unmarshal(b, &j)
+	if err != nil {
+		return Action{}, fmt.Errorf("error unmarshal: %s", err)
+
+	}
+	if len(j) != 1 {
+		return Action{}, fmt.Errorf("error, number of keys off: %d should be 1", len(j))
+	}
+
+	for action, data := range j {
+		return Action{
+			Action: action,
+			Meta:   data,
+		}, nil
+	}
+
+	panic("unreachable")
+}
+
+func (h *APIHandler) updateMetrics(action string, actionStatus int, attrs metric.MeasurementOption) bool {
+	var isErr bool
+
+	switch action {
+	case "index":
+		h.metrics.bulkIndexTotalMetrics.Add(context.Background(), 1, attrs)
+	case "create":
+		switch actionStatus {
+		case http.StatusOK:
+			h.metrics.bulkCreateOkMetrics.Add(context.Background(), 1, attrs)
+		case http.StatusConflict:
+			isErr = true
+			h.metrics.bulkCreateDuplicateMetrics.Add(context.Background(), 1, attrs)
+		case http.StatusTooManyRequests:
+			isErr = true
+			h.metrics.bulkCreateTooManyMetrics.Add(context.Background(), 1, attrs)
+		case http.StatusNotAcceptable:
+			isErr = true
+			h.metrics.bulkCreateNonIndexMetrics.Add(context.Background(), 1, attrs)
+		}
```

Comment:
@leehinman is there any reason for having the odds and the metrics only for `create` actions?